Make some functions work with Milvus

hoanq1811 opened the pull request:
**Description**
Make some functions work with Milvus:
1. get_ids: Get primary keys by field in the metadata
2. delete: Delete one or more entities by ids
3. upsert: Update/Insert one or more entities

**Issue**
None
**Dependencies**
None
**Tag maintainer:**
@hwchase17 
**Twitter handle:**
None

Commit:
Make some features to work with milvus db such as: get_pks, delete, upsert

baskaryan commented on libs/langchain/langchain/vectorstores/milvus.py:
```diff
@@ -821,3 +821,87 @@ def from_texts(
+        
+        return pks
+ 
+    def delete(self, pks: List[int] | None = None, **kwargs: Any) -> bool | None:
```
can we use the method signature inherited from parent class?
```python
    def delete(self, ids: Optional[List[str]] = None, **kwargs: Any) -> Optional[bool]:
```

hoanq1811 replied:
Yes, I’ve updated it. You can check it now

baskaryan commented:
cc @filip-halt

Commit:
Refector the formating code to correct format

hoanq1811 commented:
@baskaryan @filip-halt 
Hey, my friends, 
Please help me to review,
Thanks & Best Regards

Commit:
Change text pk by using the parameter primary_field

hoanq1811 commented:
Hi @baskaryan and @hwchase17 ,
I saw that the structure code was updated and the **delete** feature was implemented.
We have a problem: If the developers used the old version, they can't know what the primary key field (id) was
So I have updated a new feature: get_ids. That helps us easily get the primary key by expression
Meanwhile, I also added the **upsert** feature to insert or update documents

fzliu commented:
@hoanq1811 Looks great! Thanks for contributing.

fzliu commented on libs/community/langchain_community/vectorstores/milvus.py:
```diff
@@ -989,3 +989,69 @@ def _parse_document(self, data: dict) -> Document:
             metadata=data.pop(self._metadata_field) if self._metadata_field else data,
         )
+
+    def get_ids(self,
```
Minor point - I would change this to `get_pks` to avoid confusion.

hoanq1811 replied:
Hi @fzliu ,
Yup. I totally agree. So I have changed it by moving `get_ids` to `get_pks`

yiwangdr commented:
ooc, is the primary purpose to support upsert? You can call the milvus upsert api directly (https://milvus.io/docs/upsert_entities.md)

Commit:
Move get_ids to get_pks & Fixed wrong format code

Commit:
Change the description of function

hoanq1811 commented:
> ooc, is the primary purpose to support upsert? You can call the milvus upsert api directly (https://milvus.io/docs/upsert_entities.md) 

Hi @yiwangdr 
Thank for your suggestion.

Actually, I made 3 features: delete, get_pks, upsert before. But now in the latest version, I found that the `delete` feature has been integrated, so I updated it again
 
For the `upsert` feature, last time I saw it wasn't supported yet. But now, it's supported, so I will look into it for integration with the langchain.

hoanq1811 committed:
Merge branch 'master' into feature/add-delete-and-upsert-milvus

Commit:
fmt

Commit:
Merge branch 'master' into feature/add-delete-and-upsert-milvus

Commit:
cr
